File: Java/SlidingWindow/0643-maximum-average-subarray-i.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        if (nums.length==0) {
            return 0;
        }

        if (nums.length < k) {
            return 0;
        }

        if (k==0) {
            return 0;
        }

        if (nums.length==1) {
            return nums[0];
        }

        int[] prefixSum = new int[nums.length];
        prefixSum[0] = nums[0];

        for (int i=1; i<=nums.length-1; i++) {
            prefixSum[i] = prefixSum[i-1] + nums[i];
        }

        double maxAverage = -1000000.00;

        for (int i=k-1; i<=nums.length-1; i++) {
            double sumOfContiguousArray = prefixSum[i] - (i<k ? 0 : prefixSum[i-k]);
            double average = sumOfContiguousArray/k;
            if (average > maxAverage) {
                maxAverage = average;
            }
        }
        return maxAverage;
    }
}

File: Swift/SlidingWindow/0643-maximum-average-subarray-i.swift
class Solution {
    func findMaxAverage(_ nums: [Int], _ k: Int) -> Double {
        if (k == 0 || nums.isEmpty) {
		    return 0
        }
        
        if (k > nums.count) {
            return 0
        }

        var prefixArray : [Int] = []
        
        for i in 0...nums.count-1 {
            var prefixSum = nums[i] + ((i-1 >= 0) ? prefixArray[i-1] : 0)
            prefixArray.append(prefixSum)
        }

        var wS = -1
        var wE = k - 1
        var maxAvg = Double(Int.min)
        
        while (wE <= prefixArray.count-1) {
            var  windowSum = prefixArray[wE] - (wS >= 0 ? prefixArray[wS] : 0)
            var windowAvg : Double = Double(Double(windowSum)/Double(k))
            if (windowAvg > maxAvg) {
                maxAvg = windowAvg
            }
            wS += 1
            wE += 1
        }
        return maxAvg
    }
}